tokio/src/sync/notify.rs

```diff
@@ -540,6 +567,8 @@ impl Notify {
                 }
             }
 
+            // Release the lock before notifying.
+            // `guarded_list` is no longer used.
             drop(waiters);
 
             wakers.wake_all();
```

Comment:
We must clean up the linked list even if this call panics.

Reply:
Good call.

Reply:
That's a good catch. I moved the list to a new struct, which makes sure the list is cleaned up on drop.